data/src/main/java/com/microsoft/azure/kusto/data/KustoType.java

```diff
@@ -0,0 +1,55 @@
+package com.microsoft.azure.kusto.data;
+
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.util.UUID;
+import java.util.function.Function;
+
+/**
+ * Represents various ADX data types and their bindings to Java classes
+ * 
+ * @author alex
+ *
+ * @see https://docs.microsoft.com/en-us/azure/data-explorer/kusto/query/scalar-data-types/
+ *
+ * @param <C>
+ */
+public class KustoType<C> {
+	
+	public static final KustoType<Integer>			INTEGER						= new KustoType<>(Integer.class);
+	// Longs often return as "Integer", so we need to cast to "number", and then get the long value:
+	public static final KustoType<Long>				LONG						= new KustoType<>(Long.class,
+			o -> Long.valueOf(Number.class.cast(o).longValue()));
+	public static final KustoType<UUID>				GUID_UUID					= new KustoType<>(UUID.class, o -> UUID.fromString(o.toString()));
+	public static final KustoType<String>			GUID_STRING					= new KustoType<>(String.class);
+	public static final KustoType<String>			STRING						= new KustoType<>(String.class);
+	public static final KustoType<Double>			REAL_DOUBLE					= new KustoType<>(Double.class);
+	public static final KustoType<Float>			REAL_FLOAT					= new KustoType<>(Float.class,
+			o -> Float.valueOf(Double.class.cast(o).floatValue()));
+	
+	public static final KustoType<ZonedDateTime>	DATETIME_ZONED_DATE_TIME	= new KustoType<>(ZonedDateTime.class, o -> ZonedDateTime.parse(o.toString()));
+	public static final KustoType<Instant>			DATETIME_INSTANT			= new KustoType<>(Instant.class, o -> Instant.parse(o.toString()));
+	public static final KustoType<Long>				DATETIME_LONG				= new KustoType<>(Long.class, o -> Instant.parse(o.toString()).toEpochMilli());
```

Comment:
why do we need this as part of the kusto types?
I think we should keep the list here short so that users can find easily the respective Converter for their KQL data type

Reply:
We don't _need_ this. In fact, we don't _need_ any of these classes. They're a convenience we've written internally to make parsing Kusto results more straightforward and type safe. Since our mappers needed to map datetime to several java types, we placed them all here to avoid writing the conversion code separately for every mapper. I can never remember the right way to transform `Strings` into the various data types. Also, the parsing code is conceptually part of Kusto since it depends on the exact representation of `datetime` values returned by the Json library.

Reply:
this is public not internal i think the better way to go would be to leave here just the 10 kusto types and add another class called KustoExtendedTypes with all the others 

bool 
datetime 
dynamic 
guid (and uuid)
int 
long
real (which is .Net double - dont think we need float too)  
string 
timespan
decimal

Reply:
What would be the benefit of that? I think it'll just make it less clear for users - they'd have to search in two different classes for their types instead of just one.

Reply:
i think its better then looking for them and having so many options - 
If we have a main class showing the 10 kusto types it will be easy for a new user to find the type he needs as it will be the type of the column 
other - more complex casting could be mad using the Extended class 
This is a suggestion, please say if you think its still wrong

Reply:
Everything we're talking about is a suggestion :)
Anyway, I don't see the benefit of dividing the types into two classes. It would make discovering the API using an IDE's autocomplete functionality harder. How about using a double underscore or some other "delimiter" to divide the Kusto type and the Java types?

`DATETIME__INSTANT`
`DATETIME__ZONED_DATE_TIME`

Another idea I have is to use static inner classes for the various kusto types. One inner class for `datetime`, one for `string` etc. In each class, we can have the various java types the Kusto type can be converted into. Pseudocode:

```
class KustoType<..>{
    static class DateTime<...>{
        STRING = ...
        INSTANT = ...
        ZONED_DATE_TIME = ...
    }
    static class Guid<...>{
        STRING = ...
        UUID = ....
    }
}
```

Users will then specify the type using something like `KustoType.DateTime.INSTANT`. Maybe this would make the distinction between the Kusto type and the Java type clearer, and the amount of types less overwhelming.